New Twig Manual (Partial) (#1082)
contao/docs: docs/manual/layout/templates/twig/manage.en.md

---
title: 'Manage templates'
description: 'The management of the template files.'
aliases:
    - /en/layout/templates/twig/manage-template/
weight: 20
---

> Twig templates are marked differently compared to the previous PHP templates: (`.html.twig` instead of `.html5`).
> 

Reviewer: Do we need this? Because technically they do not need any extension, we're just going to ignore anything but `.twig` in our loader.

```suggestion
```

---
title: 'Manage templates'
description: 'The management of the template files.'
aliases:
    - /en/layout/templates/twig/manage-template/
weight: 20
---

Twig templates are marked differently compared to the previous PHP templates: (`.html.twig` instead of `.html5`).

{{% notice tip %}}
Contao 5 includes ready-made Twig templates and these are favored. However, you can also activate and use the corresponding PHP-templates. 
More information can be found [here](https://github.com/contao/contao/blob/5.x/UPGRADE.md#content-elements).
{{% /notice %}}

> 

Reviewer: I know your intention, but I think I'd also omit this block, specifically the part about switching back to PHP templates. If you are using Contao 5, you should use Twig templates where available. PHP templates are deprecated and the support will be removed in Contao 6, so there isn't really an option. :smile:

Author: I get it. But mentioning it may help to prevent further inquiries in this regard.
(The possibility may help when switching to Contao 5).

Reviewer: Then what about this? 

```suggestion
{{% notice tip %}}
If you are upgrading to Contao 5, make sure to check our [upgrade guide](https://github.com/contao/contao/blob/5.x/UPGRADE.md#content-elements).
Lots of elements have been rewritten and now feature new templates, but there is a way to gradually enable these to give you more time to refactor things.
{{% /notice %}}

```

## Create Template

{{< tabs groupId="creation">}}

{{% tab name="Ab Contao 4.13.10" %}}

> Ready-made Twig Temlates are not available in this version and cannot be created via the backend. You have to create a file manually 
> in the `templates` directory, e.g. `ce_text.html.twig`. This Twig template will then be displayed in the backend 
> (could be copied and renamed here) and is then available for selection in the respective content element.
> 
> It is possible to extend existing PHP-template from within Twig templates. For example, create a `fe_page.html.twig` in your 
> template directory. In this example, a heading is added above the main section and everything else stays the same:

Reviewer: I don't think it's most important to denote which templates the core is shipping (this could be added in an info box for instance) but *that* you can overwrite/extend any template:

```suggestion
You can now use Twig templates to overwrite or extend PHP templates. Keep the file name like before, but adjust the file extension. For example, create a `fe_page.html.twig` in your `template` directory and add another heading above the main section:
```

Author: Done

```twig
{# /templates/fe_page.html.twig #}

{% extends '@Contao/fe_page' %}

{% block main %}
  <h1>Hello from Twig!</h1>
  {{ parent() }}
{% endblock %}
```

{{% notice info %}}
There can be either **one** Twig **or** a PHP variant of the same template in the same place. You cannot extend Twig templates 
from PHP templates, only vice versa.
{{% /notice %}}

{{% /tab %}}

{{% tab name="Ab Contao 5.0.2" %}}

> This version includes ready-made Twig templates and these can be selected and created via the backend for customization. 
> The corresponding templates are expanded by default and the respective sections can then be overwritten. 
> You can also add your own code here. 

Reviewer: ```suggestion
Beginning in Contao 5.0, we started replacing our core templates with Twig templates. Besides having new features, these new templates are also making use of the new directory structure that you need to follow when overwriting/extending - e.g. `content_element/gallery.html.twig` means a `gallery.html.twig` file inside the `content_element` directory.

You can now also derive from existing Twig templates in the 'Layout > Templates' section in the back end. Then we'll automatically add an `extends` statement for the base template along with some explaining comments into the newly created file. You should favor extending, i.e. adjusting only the needed parts, over overwriting in nearly all cases - this way you'll stay compatible with future changes and features that are introduced by extensions.
```

Author: Done

An example for the content element "Text":

```twig
{% extends "@Contao/content_element/text.html.twig" %}

{#
  ** Add changes to the base template here. **

  Hint: Try adjusting blocks and attributes instead of
  overwriting the whole template. This way your version
  can remain compatible with future changes to the base
  template as well as adjustments made by extensions.

  Currently available blocks:
    "picture_component", "image", "sources", "source",
    "schema_org", "figure_component", "media",
    "media_link", "caption", "caption_inner",
    "content", "text_media", "text",
    "text_attributes", "headline_component",
    "headline_attributes", "headline_inner",
    "wrapper", "wrapper_tag", "attributes", "inner"

  Example:
    {% block picture_component %}
       {{ parent() }}
       My additional content for 'picture_component'…
    {% endblock %}

> #}

Reviewer: We should probably add a real example here instead of the auto-generated comment.

Author: Done

```

{{% notice note %}}
A Twig template may not yet be available for each module/content element. In these cases the previous previous (PHP/legacy) templates will be used.
{{% /notice %}}

{{% /tab %}}

{{< /tabs >}}

## Template Variant

{{< tabs groupId="variant">}}

{{% tab name="Ab Contao 4.13.10" %}}

The template files are prefixed for recognition. For example, `ce_` indicates a content element (**c**ontent **e**lement). For example, 
if you want to change the output of the content element of type "text", you can use the template `ce_text.html.twig` for this purpose. 
In this case the template changes will affect all content elements of type "Text". 

This is not always desired. For specific use, the template can be named individually. Here the 
template name must be kept and only extended: 

You can rename e.g. `ce_text.html.twig` to `ce_text_individual.html.twig`. This template can then be used specifically for the output 
for one (or more) content element(s) of type "Text".

{{% /tab %}}

{{% tab name="Ab Contao 5.0.2" %}}

> If you want to offer one or more variants of a template, which can be selected in the backend, then create a subfolder analogous to 
> to the file name of the template and place the customized file in it. 
> 
> Example: You can put a text variant `highlight` under `content_element/text/highlight.html.twig`. 
> This is then available as an individual template `content_element/text/highlight` for selection in the content element. 

Reviewer: ```suggestion
If you want to offer one or more variants of a template, which can be selected in the back end, then create a sub directory analogous to the file name of the template and place the customized file in it. 

Example: You can put a text variant `highlight` under `content_element/text/highlight.html.twig`. 
You can then select `content_element/text/highlight` as "Individual template" in the "Text" content element. 
```

Author: Done
